packages/preferences/src/browser/folders-preferences-provider.ts

```diff
@@ -133,45 +133,59 @@ export class FoldersPreferencesProvider extends PreferenceProvider {
     }
 
     async setPreference(preferenceName: string, value: any, resourceUri?: string): Promise<boolean> {
-        const sectionName = preferenceName.split('.', 1)[0];
-        const configName = this.configurations.isSectionName(sectionName) ? sectionName : this.configurations.getConfigName();
+        const firstPathFragment = preferenceName.split('.', 1)[0];
```

Comment:
A lot of code has been rewritten here. The substantive change is that after trying the files matching the section name (e.g. `extensions.json` if the preference starts with `'extensions'`), we fall back to `settings.json` if the change has been rejected. This allows the provider for `extensions.json` to reject changes that contain the prefix `extensions` but are not in the schema for that file (e.g. `extensions.ignoreRecommendations`).

There is also a stylistic change in the setup of the final iterators. The code on `master` uses an array containing functions that *either* return a provider *or* add another function to the end of the array. The conditions for returning the provider are increasingly broad with each subsequent function, so it was effectively a prioritization mechanism that avoided a `sort` call.

The new code follows a similar principle, but instead of gradually adding to the same array, it creates a number of prioritized buckets, guaranteeing that we only consider each candidate once after it has been prioritized, and, I hope, clarifying the precedence order of the various conditions.